src/interpreter/ByteCodeInterpreter.cpp

```diff
@@ -3047,20 +3047,18 @@ NEVER_INLINE void ByteCodeInterpreter::objectDefineOwnPropertyOperation(Executio
         value.asFunction()->defineOwnProperty(state, state.context()->staticStrings().name, ObjectPropertyDescriptor(fnName));
     }
 
-    // http://www.ecma-international.org/ecma-262/6.0/#sec-__proto__-property-names-in-object-initializers
-    if (!willBeObject.asObject()->isScriptClassConstructorPrototypeObject() && (propertyStringOrSymbol.isString() && propertyStringOrSymbol.asString()->equals("__proto__"))) {
-        willBeObject.asObject()->setPrototype(state, value);
-    } else {
-        willBeObject.asObject()->defineOwnProperty(state, ObjectPropertyName(state, propertyStringOrSymbol), ObjectPropertyDescriptor(value, code->m_presentAttribute));
-    }
+    willBeObject.asObject()->defineOwnProperty(state, ObjectPropertyName(state, propertyStringOrSymbol), ObjectPropertyDescriptor(value, code->m_presentAttribute));
 }
 
 NEVER_INLINE void ByteCodeInterpreter::objectDefineOwnPropertyWithNameOperation(ExecutionState& state, ObjectDefineOwnPropertyWithNameOperation* code, Value* registerFile)
 {
     const Value& willBeObject = registerFile[code->m_objectRegisterIndex];
     // http://www.ecma-international.org/ecma-262/6.0/#sec-__proto__-property-names-in-object-initializers
     if (!willBeObject.asObject()->isScriptClassConstructorPrototypeObject() && (code->m_propertyName == state.context()->staticStrings().__proto__)) {
-        willBeObject.asObject()->setPrototype(state, registerFile[code->m_loadRegisterIndex]);
+        const Value v = registerFile[code->m_loadRegisterIndex];
+        if (v.isObject() || v.isNull()) {
+            willBeObject.asObject()->setPrototype(state, registerFile[code->m_loadRegisterIndex]);
```

Comment:
How about passing `v` value for argument of `setPrototype` instead of accessing `registerFile[code->m_loadRegisterIndex]` again at this?

Reply:
Good catch